Kernel/Graphics/VirtIOGPU/GraphicsAdapter.cpp

```diff
@@ -421,10 +416,22 @@ Graphics::VirtIOGPU::ResourceID VirtIOGraphicsAdapter::allocate_resource_id()
     return m_resource_id_counter++;
 }
 
-Graphics::VirtIOGPU::ContextID VirtIOGraphicsAdapter::allocate_context_id()
+ErrorOr<Graphics::VirtIOGPU::ContextID> VirtIOGraphicsAdapter::allocate_context_id()
 {
-    // FIXME: This should really be tracked using a bitmap, instead of an atomic counter
-    return m_context_id_counter++;
+    Optional<Graphics::VirtIOGPU::ContextID> new_context_id;
+
+    m_active_context_ids.with([&new_context_id](Bitmap& active_context_ids) {
+        auto maybe_available_id = active_context_ids.find_first_unset();
+        if (!maybe_available_id.has_value())
+            return;
+        new_context_id = maybe_available_id.value();
+        active_context_ids.set(maybe_available_id.value(), true);
+    });
+
+    if (new_context_id.has_value())
+        return new_context_id.value();
+
+    return Error::from_string_literal("VirtIO::GraphicsAdapter: No available context IDs.");
```

Comment:
I am not sure this is actually the appropriate thing to do. I looked where we use `from_string_literal()` in the kernel and we should not use it. At least the usage in `NetworkingManagement` looks a bit OK (it was added by @awesomekling recently), but we should ideally propagate errors with the errno code if it will propagate all the way back to userspace.

Reply:
Any suggestions of what errno code it should be? I had a look and none of them seemed right.

Reply:
Either `ENXIO` or `ERANGE` seem OK (I don't really have strong preference about either of these, but `ENXIO` seem a bit more sensible here, so feel free to choose here).

Reply:
Thanks!